File: src/go/pkg/netipc/service/raw/apps_lookup.go
package raw

import (
	"errors"

	"github.com/netdata/netdata/go/plugins/pkg/netipc/protocol"
)

// AppsLookupHandler serves a single APPS_LOOKUP service kind.
type AppsLookupHandler func(*protocol.AppsLookupRequestView, *protocol.AppsLookupBuilder) bool

func appsLookupRequestSize(pids []uint32) (int, error) {
	return appsLookupRequestSizeForCount(len(pids))
}

func appsLookupRequestSizeForCount(count int) (int, error) {
	dirSize, err := checkedLookupMul(count, protocol.LookupDirEntrySize)
	if err != nil {
		return 0, err
	}
	keySize, err := checkedLookupMul(count, protocol.AppsLookupKeySize)
	if err != nil {
		return 0, err
	}
	size, err := checkedLookupAdd(protocol.AppsLookupReqHdr, dirSize)
	if err != nil {
		return 0, err
	}
	return checkedLookupAdd(size, keySize)
}

func appsLookupNextRequest(pids []uint32, maxPayload uint32) (int, int, error) {
	if len(pids) == 0 {
		size, err := appsLookupRequestSize(nil)
		return 0, size, err
	}
	if maxPayload == 0 {
		maxPayload = protocol.MaxPayloadDefault
	}
	perItem := protocol.LookupDirEntrySize + protocol.AppsLookupKeySize
	if maxPayload < uint32(protocol.AppsLookupReqHdr+perItem) {
		return 0, 0, protocol.ErrOverflow
	}
	maxCount := int(maxPayload-uint32(protocol.AppsLookupReqHdr)) / perItem
	if maxCount <= 0 {
		return 0, 0, protocol.ErrOverflow
	}
	count := len(pids)
	if count > maxCount {
		count = maxCount
	}
	size, err := appsLookupRequestSize(pids[:count])
	if err != nil {
		return 0, 0, err
	}
	return count, size, nil
}

// CallAppsLookup performs a blocking typed APPS_LOOKUP call.
// The returned view is valid until the next typed call on this client.
func (c *Client) CallAppsLookup(pids []uint32) (*protocol.AppsLookupResponseView, error) {
	return c.CallAppsLookupWithTimeout(pids, 0)
}

func (c *Client) CallAppsLookupWithTimeout(pids []uint32, timeoutMs uint32) (*protocol.AppsLookupResponseView, error) {
	if err := c.validateMethod(protocol.MethodAppsLookup); err != nil {
		return nil, err
	}
	if uint64(len(pids)) > uint64(c.maxLogicalLookupItems) {
		return nil, protocol.ErrOverflow
	}
	if err := c.ensureReadyForLogicalLookup(); err != nil {
		return nil, err
	}

	expectedCount, err := checkedLookupU32(len(pids))
	if err != nil {
		return nil, err
	}
	rawItems := make([][]byte, 0, len(pids))
	start := 0
	var generation uint64
	haveGeneration := false
	subcalls := uint32(0)
	deadline := c.newLookupDeadline(timeoutMs)
	for {
		reqCount, reqSize, err := appsLookupNextRequest(pids[start:], c.sessionMaxRequestPayloadBytes())
		if err != nil {
			oneItemSize, serr := appsLookupRequestSize([]uint32{0})
			if errors.Is(err, protocol.ErrOverflow) && start < len(pids) && serr == nil {
				if cerr := c.ensureLookupRequestCapacity(oneItemSize); cerr == nil {
					continue
				}
			}
			return nil, err
		}
		reqPids := pids[start : start+reqCount]
		reqBuf := ensureClientScratch(&c.requestBuf, reqSize)
		reqLen, err := protocol.EncodeAppsLookupRequest(reqPids, reqBuf)
		if err != nil {
			return nil, err
		}

		remainingTimeout, terr := c.lookupRemainingTimeout(deadline)
		if terr != nil {
			return nil, terr
		}
		subcalls++
		if subcalls > c.maxLogicalLookupSubcalls {
			return nil, protocol.ErrOverflow
		}
		var payload []byte
		if err := c.callWithRetry(func() error {
			_, callPayload, rerr := c.doRawCallWithTimeout(protocol.MethodAppsLookup, reqBuf[:reqLen], remainingTimeout)
			if rerr != nil {
				return rerr
			}
			payload = callPayload
			return nil
		}); err != nil {
			return nil, err
		}
		view, derr := protocol.DecodeAppsLookupResponse(payload)
		if derr != nil {
			return nil, derr
		}
		if !haveGeneration {
			generation = view.Generation
			haveGeneration = true
		} else if view.Generation != generation {
			return nil, protocol.ErrBadLayout
		}
		reqCountU32, err := checkedLookupU32(len(reqPids))
		if err != nil {
			return nil, err
		}
		if view.ItemCount != reqCountU32 {
			return nil, protocol.ErrBadItemCount
		}
		payloadExceededAt := -1
		for i, expected := range reqPids {
			itemIndex, ierr := checkedLookupU32(i)
			if ierr != nil {
				return nil, ierr
			}
			item, ierr := view.Item(itemIndex)
			if ierr != nil {
				return nil, ierr
			}
			if item.Pid != expected {
				return nil, protocol.ErrBadLayout
			}
			if item.Status == protocol.PidLookupPayloadExceeded {
				payloadExceededAt = i
				break
			}
			rawItem, ierr := view.RawItem(itemIndex)
			if ierr != nil {
				return nil, ierr
			}
			rawItems = append(rawItems, cloneLookupRawItem(rawItem))
		}
		if payloadExceededAt < 0 {
			start += reqCount
			if start < len(pids) {
				continue
			}
			rawItemCount, cerr := checkedLookupU32(len(rawItems))
			if cerr != nil {
				return nil, cerr
			}
			if rawItemCount != expectedCount {
				return nil, protocol.ErrBadItemCount
			}
			size, serr := lookupRawResponseSize(protocol.AppsLookupRespHdr, rawItems)
			if serr != nil {
				return nil, serr
			}
			if lookupSizeExceedsLimit(size, c.maxLogicalLookupResponseBytes) {
				return nil, protocol.ErrOverflow
			}
			stitched := make([]byte, size)
			n, eerr := protocol.EncodeAppsLookupRawResponse(rawItems, generation, stitched)
			if eerr != nil {
				return nil, eerr
			}
			return protocol.DecodeAppsLookupResponse(stitched[:n])
		}
		for i := payloadExceededAt; i < len(reqPids); i++ {
			itemIndex, ierr := checkedLookupU32(i)
			if ierr != nil {
				return nil, ierr
			}
			item, ierr := view.Item(itemIndex)
			if ierr != nil {
				return nil, ierr
			}
			if item.Pid != reqPids[i] || item.Status != protocol.PidLookupPayloadExceeded {
				return nil, protocol.ErrBadLayout
			}
		}
		if payloadExceededAt == 0 {
			return nil, protocol.ErrOverflow
		}
		start += payloadExceededAt
	}
}

// AppsLookupDispatch adapts a typed apps lookup handler to the raw dispatch shape.
func AppsLookupDispatch(handle AppsLookupHandler) DispatchHandler {
	if handle == nil {
		return nil
	}
	return func(request []byte, responseBuf []byte) (int, error) {
		n, err := protocol.DispatchAppsLookup(request, responseBuf, func(req *protocol.AppsLookupRequestView, builder *protocol.AppsLookupBuilder) bool {
			return handle(req, builder)
		})
		if errors.Is(err, protocol.ErrHandlerFailed) {
			return 0, errHandlerFailed
		}
		return n, err
	}
}

File: src/crates/netipc/src/service/raw/apps_lookup.rs
use super::client::{ClientConfig, RawCallKind, RawClient};
use super::common::lookup_raw_response_size;
use super::dispatch::{dispatch_error_from_protocol, DispatchHandler};
use crate::protocol::{
    self, AppsLookupBuilder, AppsLookupRequestView, AppsLookupResponseView, NipcError,
    APPS_LOOKUP_KEY_SIZE, APPS_LOOKUP_REQ_HDR_SIZE, APPS_LOOKUP_RESP_HDR_SIZE,
    LOOKUP_DIR_ENTRY_SIZE, METHOD_APPS_LOOKUP, PID_LOOKUP_PAYLOAD_EXCEEDED,
};
use std::sync::Arc;

fn apps_lookup_request_size(count: usize) -> Result<usize, NipcError> {
    let dir_size = count
        .checked_mul(LOOKUP_DIR_ENTRY_SIZE)
        .ok_or(NipcError::Overflow)?;
    let key_size = count
        .checked_mul(APPS_LOOKUP_KEY_SIZE)
        .ok_or(NipcError::Overflow)?;
    APPS_LOOKUP_REQ_HDR_SIZE
        .checked_add(dir_size)
        .and_then(|v| v.checked_add(key_size))
        .ok_or(NipcError::Overflow)
}

fn apps_lookup_next_request_count(
    remaining: usize,
    max_payload: u32,
) -> Result<(usize, usize), NipcError> {
    if remaining == 0 {
        return Ok((0, apps_lookup_request_size(0)?));
    }

    let cap = if max_payload == 0 {
        protocol::MAX_PAYLOAD_DEFAULT
    } else {
        max_payload
    } as usize;
    let per_item = LOOKUP_DIR_ENTRY_SIZE
        .checked_add(APPS_LOOKUP_KEY_SIZE)
        .ok_or(NipcError::Overflow)?;
    if cap < APPS_LOOKUP_REQ_HDR_SIZE || cap - APPS_LOOKUP_REQ_HDR_SIZE < per_item {
        return Err(NipcError::Overflow);
    }
    let max_count = (cap - APPS_LOOKUP_REQ_HDR_SIZE) / per_item;
    if max_count == 0 {
        return Err(NipcError::Overflow);
    }
    let count = remaining.min(max_count);
    Ok((count, apps_lookup_request_size(count)?))
}

pub type AppsLookupHandler =
    Arc<dyn for<'a> Fn(&AppsLookupRequestView, &mut AppsLookupBuilder<'a>) -> bool + Send + Sync>;

impl RawClient {
    /// Create a new client context bound to the apps-lookup service kind.
    /// Does NOT connect. Does NOT require the server to be running.
    pub fn new_apps_lookup(run_dir: &str, service_name: &str, config: ClientConfig) -> Self {
        Self::new_bound(run_dir, service_name, METHOD_APPS_LOOKUP, config)
    }

    /// Blocking typed call: APPS_LOOKUP method.
    ///
    /// The returned view is valid until the next typed call on this client.
    pub fn call_apps_lookup(
        &mut self,
        pids: &[u32],
    ) -> Result<AppsLookupResponseView<'_>, NipcError> {
        self.call_apps_lookup_with_timeout(pids, 0)
    }

    /// Blocking typed call with an explicit timeout in milliseconds.
    ///
    /// A zero timeout uses the client's context-level default.
    pub fn call_apps_lookup_with_timeout(
        &mut self,
        pids: &[u32],
        timeout_ms: u32,
    ) -> Result<AppsLookupResponseView<'_>, NipcError> {
        self.validate_method(METHOD_APPS_LOOKUP)?;
        if pids.len() > self.max_logical_lookup_items as usize {
            return Err(NipcError::Overflow);
        }
        self.ensure_ready_for_logical_lookup()?;

        let mut raw_items: Vec<Vec<u8>> = Vec::with_capacity(pids.len());
        let mut start = 0usize;
        let mut generation: Option<u64> = None;
        let mut subcalls = 0u32;
        let deadline = self.lookup_deadline(timeout_ms)?;
        loop {
            let (req_count, req_size) = match apps_lookup_next_request_count(
                pids.len() - start,
                self.session_max_request_payload_bytes(),
            ) {
                Ok(next) => next,
                Err(NipcError::Overflow) if start < pids.len() => {
                    let one_item_size = apps_lookup_request_size(1)?;
                    if self.ensure_lookup_request_capacity(one_item_size)? {
                        continue;
                    }
                    return Err(NipcError::Overflow);
                }
                Err(err) => return Err(err),
            };
            let req_pids = &pids[start..start + req_count];
            let req_len = {
                let req_buf = self.request_scratch(req_size);
                protocol::encode_apps_lookup_request(req_pids, req_buf)?
            };
            let remaining_timeout = self.lookup_remaining_timeout(deadline)?;
            subcalls += 1;
            if subcalls > self.max_logical_lookup_subcalls {
                return Err(NipcError::Overflow);
            }
            let response = self.raw_call_with_retry_timeout(
                METHOD_APPS_LOOKUP,
                req_len,
                RawCallKind::single(),
                remaining_timeout,
            )?;
            let final_size = {
                let view = AppsLookupResponseView::decode(self.response_payload(response)?)?;
                if let Some(expected_generation) = generation {
                    if view.generation != expected_generation {
                        return Err(NipcError::BadLayout);
                    }
                } else {
                    generation = Some(view.generation);
                }
                if view.item_count != req_pids.len() as u32 {
                    return Err(NipcError::BadItemCount);
                }
                let mut payload_exceeded_at: Option<usize> = None;
                for (i, expected) in req_pids.iter().enumerate() {
                    let item = view.item(i as u32)?;
                    if item.pid != *expected {
                        return Err(NipcError::BadLayout);
                    }
                    if item.status == PID_LOOKUP_PAYLOAD_EXCEEDED {
                        payload_exceeded_at = Some(i);
                        break;
                    }
                    raw_items.push(view.raw_item(i as u32)?.to_vec());
                }
                if let Some(first) = payload_exceeded_at {
                    for (i, expected) in req_pids.iter().enumerate().skip(first) {
                        let item = view.item(i as u32)?;
                        if item.pid != *expected || item.status != PID_LOOKUP_PAYLOAD_EXCEEDED {
                            return Err(NipcError::BadLayout);
                        }
                    }
                    if first == 0 {
                        return Err(NipcError::Overflow);
                    }
                    start += first;
                    continue;
                }
                start += req_count;
                if start < pids.len() {
                    continue;
                }
                if raw_items.len() != pids.len() {
                    return Err(NipcError::BadItemCount);
                }
                let size = lookup_raw_response_size(
                    APPS_LOOKUP_RESP_HDR_SIZE,
                    raw_items.len(),
                    &raw_items,
                )?;
                if size > self.max_logical_lookup_response_bytes as usize {
                    return Err(NipcError::Overflow);
                }
                size
            };
            self.transport_buf.resize(final_size, 0);
            let refs: Vec<&[u8]> = raw_items.iter().map(Vec::as_slice).collect();
            let n = protocol::encode_apps_lookup_raw_response(
                &refs,
                generation.unwrap_or(0),
                &mut self.transport_buf[..final_size],
            )?;
            return AppsLookupResponseView::decode(&self.transport_buf[..n]);
        }
    }
}

pub fn apps_lookup_dispatch(handler: AppsLookupHandler) -> DispatchHandler {
    Arc::new(move |request, response_buf| {
        protocol::dispatch_apps_lookup(request, response_buf, |request, builder| {
            handler(request, builder)
        })
        .map_err(dispatch_error_from_protocol)
    })
}
